alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -623,3 +644,139 @@ instance
   FromCBOR (UtxoPredicateFailure era)
   where
   fromCBOR = decode (Summands "UtxoPredicateFailure" decFail)
+
+data ScriptFailure c
+  = MissingRedeemer RdmrPtr
+  | InvalidScriptPurpose (Alonzo.ScriptPurpose c)
+  | MissingScript (ScriptHash c)
+  | MissingDatum (Alonzo.DataHash c)
+  | ValidationFailed P.EvaluationError
+  | UnknownTxIn (TxIn c)
+  | InvalidTxIn (TxIn c)
+  | IncompatibleBudget P.ExBudget
+
+evaluateTransactionExecutionUnits ::
+  forall era.
+  ( Era era,
+    Core.TxOut era ~ TxOut era,
+    Core.Script era ~ Script era,
+    Core.Value era ~ Mary.Value (Crypto era),
+    Core.Witnesses era ~ TxWitness era,
+    Core.TxBody era ~ Alonzo.TxBody era
+  ) =>
+  Core.Tx era ->
+  UTxO era ->
+  EpochInfo Identity ->
+  SystemStart ->
+  CostModel ->
+  Map
+    (ScriptHash (Crypto era))
+    (Either (ScriptFailure (Crypto era)) (RdmrPtr, ExUnits))
+evaluateTransactionExecutionUnits tx utxo ei sysS (CostModel costModel) =
+  Map.fromList $ map findAndCount neededPlutusScripts
+  where
+    note :: e -> Maybe a -> Either e a
+    note _ (Just x) = Right x
+    note e Nothing = Left e
+
+    txb = getField @"body" tx
+    ws = getField @"wits" tx
+    rs = unRedeemers $ getField @"txrdmrs" ws
+    dats = unTxDats $ getField @"txdats" ws
+    scripts = getField @"txscripts" ws
+    sNeeded = scriptsNeeded utxo tx
+    neededPlutusScripts = do
+      (sp, sh) <- sNeeded
+      msb <- case Map.lookup sh scripts of
+        Nothing -> pure Nothing
+        Just (TimelockScript _) -> []
+        Just (PlutusScript bytes) -> pure $ Just bytes
+      pure (sp, sh, msb)
+
+    txinfo = txInfo ei sysS utxo tx
+
+    exBudgetToExUnits :: P.ExBudget -> Maybe ExUnits
+    exBudgetToExUnits (P.ExBudget (P.ExCPU cpu) (P.ExMemory memory)) =
+      ExUnits <$> convertCost cpu <*> convertCost memory
+
+    convertCost :: P.CostingInteger -> Maybe Word64
```

Comment:
@michaelpj do you know if there is a better way for me to make this conversion? (for context, this is a helper function that will be used in the Cardano API (the one in node) and not in the ledger rules.

Reply:
I think this is more generic, right? You basically want:
```
safeFromIntegral :: (Integral a, Num b, Bounded b) => a -> Maybe b
```
Oddly I can't seem to find this function anywhere.

Reply:
I think what you looking for is `Integral`, not `Num`. I'd also call it explicitly `safeFromInteger` and make users of the function to call `toInteger`, this way it is clear how expensive the function is.

```haskell
safeFromInteger :: forall a. (Integral a, Bounded a) => Integer -> Maybe a
safeFromInteger i
  | toInteger (minBound :: a) <= i && i <= toInteger (maxBound :: a) = Just $ fromInteger i
  | otherwise = Nothing
```

Reply:
thanks y'all! the generic solution sounds great, I just wasn't sure if the plutus library already had ways of handling the `CostingInteger` alias.